software/bsg_manycore_lib/bsg_mcs_mutex.S

```diff
@@ -0,0 +1,103 @@
+//
+// Cores atomically append their local memory region to the global list using an unconditional
+// amoswap operation. They then spin on their local memories for a predecessor in the queue
+// to notify them that they now hold the lock.
+//
+// Once a core has completed its critical region, it checks for a successor and updates releases the lock to them.
+//
+// The advantages of this mutex over a simple spin lock on the manycore are two fold:
+//
+// (1) It greatly reduces the number of memory requests on the network and it mitigates the extent to which
+// a single memory bank becomes a hot-spot. The number of requests issued to a memory bank containing the
+// lock object is linear with the number of times an acquire operation is executed.
+//
+// (2) The lock approximates a FIFO-ish structure, which improves fairness. A simple spinlock on the manycore
+// will favor threads topologically closer to the memory bank in which the lock resides and can lead to
+// starvation of the other cores.
+//
+// This lock is by no means perfect. For locks with low contention, a simple spinlock may result in better performance.
+
+        .text
+        .globl bsg_mcs_mutex_acquire
+        // Refer to bsg_mcs_mutex.h for detailed description of usage.
+        // a0 = mtx         : bsg_mcs_mutex_t*, points to DRAM
+        // a1 = lcl         : bsg_mcs_mutex_node_t*, local pointer that points to DMEM
+        // a2 = lcl_as_glbl : bsg_mcs_mutex_node_t*, global pointer to same location as 'lcl'
+bsg_mcs_mutex_acquire:
+        sw      x0, 4(a1)                      // lcl->next = 0
+        sw      x0, 0(a1)                      // lcl->unlocked = 0
+        amoswap.w.aq t0, a2, 0(a0)             // predecessor = swap (&mtx, lcl_as_glbl)
+        beqz    t0, bsg_mcs_mutex_acquire_ret  // return if predecessor = 0
+        sw      a2, 4(t0)                      // predecessor->next = lcl_as_glbl
+bsg_mcs_mutex_acquire_loop:
+        // Here we use the load-on-broken-reservation semantics to avoid
+        // busy waiting. This reduces the dynamic energy of the core
+        // and removes contention on our local memory from updates by
+        // other cores, including an update from our predecessor
+        // for when they release the lock to us.
+        //
+        // The expected wait time for this load is arbitrarily long as it depends
+        // on (1) the time it takes client code to complete the critical region
+        // and (2) the contention on this lock.
+        // We expect the wait time to be on the order of 20-100 cycles in the
+        // case where there is low contention on the lock.
+        lr.w    t0, (a1)                       // unlocked = lcl->unlocked
+        bnez    t0, bsg_mcs_mutex_acquire_ret  // return if unlocked
+        lr.w.aq t0, (a1)                       // unlocked = lcl->unlocked
+        // MBT: backwards predict not taken branch variant would be helpful here
```

Comment:
Just add a few notes here should we revisit:

// MBT: if we supported context switching the reservation register, then we would not even
//         need this branch (i.e. a blind synchronization); but if there was a context switch, 
//         then the reservation would be cleared and we would this need to go back to the lr.w 
//         to reprime the barrier.
//
// MBT: if lr.w.aq re-primed the reservation AND we supported context switching the reservation
// register, then we could just jump back to the lr.w.aq instruction